kotlin/internal/jvm/compile.bzl

```diff
@@ -237,81 +267,174 @@ def kt_jvm_compile_action(ctx, rule_kind, output_jar, compile_jar):
     dirs = _compiler_directories(ctx)
     srcs = _partitioned_srcs(ctx.files.srcs)
     friend = _compiler_friends(ctx, friends = getattr(ctx.attr, "friends", []))
-    compile_deps = _compiler_deps(toolchains, friend, deps = ctx.attr.deps + ctx.attr.plugins)
+    compile_deps = _compiler_deps(toolchains, friend, deps = ctx.attr.deps)
     annotation_processors = _plugin_mappers.targets_to_annotation_processors(ctx.attr.plugins + ctx.attr.deps)
     transitive_runtime_jars = _plugin_mappers.targets_to_transitive_runtime_jars(ctx.attr.plugins + ctx.attr.deps)
     plugins = ctx.attr.plugins
 
-    if toolchains.kt.experimental_use_abi_jars:
-        kt_compile_jar = ctx.actions.declare_file(ctx.label.name + "-kt.abi.jar")
+    output_jars = []
+    compile_jars = []
+    kt_stubs_for_java = []
+    generated_src_jars = []
+
+    # Run KAPT
+    if srcs.kt and annotation_processors:
+        kapt_generated_src_jar = ctx.actions.declare_file(ctx.label.name + "-kapt-generated-src.jar")
+        kapt_generated_stub_jar = ctx.actions.declare_file(ctx.label.name + "-kapt-generated-stub.jar")
+        kapt_generated_class_jar = ctx.actions.declare_file(ctx.label.name + "-kapt-generated-class.jar")
         _run_kt_builder_action(
             ctx = ctx,
             rule_kind = rule_kind,
             toolchains = toolchains,
             dirs = dirs,
             srcs = srcs,
+            generated_src_jars = [],
             friend = friend,
             compile_deps = compile_deps,
             annotation_processors = annotation_processors,
             transitive_runtime_jars = transitive_runtime_jars,
-            plugins = plugins,
+            plugins = [],
             outputs = {
-                "abi_jar": kt_compile_jar,
+                "generated_java_srcjar": kapt_generated_src_jar,
+                "kapt_generated_stub_jar": kapt_generated_stub_jar,
+                "kapt_generated_class_jar": kapt_generated_class_jar,
             },
-            mnemonic = "KotlinCompileAbi",
+            build_java = False,
+            build_kotlin = False,
+            mnemonic = "KotlinKapt",
         )
-        if not srcs.java:
-            compile_jar = kt_compile_jar
+        generated_src_jars.append(kapt_generated_src_jar)
+        output_jars.append(kapt_generated_class_jar)
```

Comment:
Hmm, actually, I don't think we want to merge `kapt_generated_class_jar` into the outputs here.

I _believe_ this jar is the compiled KAPT generated sources (perhaps stub versions), so I can try including them on the `KotlinCompile` classpath. It should be faster than parsing the sources, and if they're actually the stubs would allow action caching if the KAPT sources don't have API changes.

 I _think_ we could possibly pass this jar on the classpath to `kotlin` rather than the generated sources. I can give this a try.

Reply:
This is a great idea, specially if you think we're going to see a visible performance improvement.

Reply:
Actually, I see what `kapt_generated_class_jar` is now, it basically contains anything that should be in the classes folder that isn't classes (since KAPT doesn't compile anything), in practice this means generated resource files, META-INF/services files etc, so we DO need to keep this, but the good thing is it doesn't include classes so won't affect performance.